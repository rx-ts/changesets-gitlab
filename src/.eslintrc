{
  "rules": {
    "sonarjs/no-nested-template-literals": 0,
    "unicorn/filename-case": 0
  }
}
