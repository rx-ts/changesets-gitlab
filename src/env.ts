import dotenv from 'dotenv'

import type { Env } from './env.types'

const prepareEnv = () => {
  dotenv.config()

  return {
    ...process.env,

    CI_MERGE_REQUEST_IID: +process.env.CI_MERGE_REQUEST_IID!,
    GITLAB_CI_USER_EMAIL:
      process.env.GITLAB_CI_USER_EMAIL ??
      'gitlab[bot]@users.noreply.gitlab.com',
    GITLAB_COMMENT_TYPE: process.env.GITLAB_COMMENT_TYPE ?? 'discussion',
    DEBUG_GITLAB_CREDENTIAL: process.env.DEBUG_GITLAB_CREDENTIAL ?? 'false',

    // only check for the token if we are explicitly using it
    // eslint-disable-next-line sonar/function-name
    get GITLAB_TOKEN() {
      if (!process.env.GITLAB_TOKEN) {
        throw new Error(
          'Please add the `GITLAB_TOKEN` to the changesets action',
        )
      }
      return process.env.GITLAB_TOKEN
    },
  } as Env
}

export const env = prepareEnv()
