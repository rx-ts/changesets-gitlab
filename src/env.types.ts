export type Env = GitLabCIPredefinedVariables &
  MergeRequestVariables & {
    GITLAB_HOST: string

    GITLAB_TOKEN: string
    GITLAB_TOKEN_TYPE: LooseString<'job' | 'oauth'>
    GITLAB_CI_USER_NAME: string | undefined
    GITLAB_CI_USER_EMAIL: string
    GITLAB_COMMENT_TYPE: LooseString<'discussion' | 'note'>
    DEBUG_GITLAB_CREDENTIAL: LooseString<'1' | 'true'>

    HOME: string
    NPM_TOKEN: string | undefined
  }

type MergeRequestVariables =
  | {
      // this is used to be checked if the current job is a merge request job
      CI_MERGE_REQUEST_SOURCE_BRANCH_NAME: undefined
    }
  | {
      CI_MERGE_REQUEST_IID: number
      CI_MERGE_REQUEST_PROJECT_URL: string
      CI_MERGE_REQUEST_SOURCE_BRANCH_NAME: string
      CI_MERGE_REQUEST_SOURCE_BRANCH_SHA: string
      CI_MERGE_REQUEST_TITLE: string
    }

type GitLabCIPredefinedVariables = { GITLAB_USER_NAME: string } & (
  | {
      // this is used to be checked if the current job is a merge request job
      CI: undefined
    }
  | {
      CI: 'true'
      CI_PROJECT_PATH: string
      CI_SERVER_URL: string
    }
)

// This type represents a couple of possible literal values for a string, but also
// allows any other string value. It's useful for environment variables that can
// have a default value, but can also be overridden by the user.
// The weird `string & {}` is to make sure that the type is not narrowed to `string`
// See: https://twitter.com/mattpocockuk/status/1671908303918473217
// eslint-disable-next-line @typescript-eslint/ban-types, sonar/no-useless-intersection
type LooseString<T extends string> = T | (string & {})
