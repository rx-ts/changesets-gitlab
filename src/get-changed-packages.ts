import nodePath from 'node:path'

import _assembleReleasePlan from '@changesets/assemble-release-plan'
import { parse as parseConfig } from '@changesets/config'
import _parseChangeset from '@changesets/parse'
import type {
  PackageJSON,
  PreState,
  NewChangeset,
  WrittenConfig,
} from '@changesets/types'
import type { Gitlab } from '@gitbeaker/core'
import type { Packages, Tool } from '@manypkg/get-packages'
import fs from 'fs-extra'
import { load } from 'js-yaml'
import micromatch from 'micromatch'

import { getAllFiles } from './utils.js'

// workaround for https://github.com/atlassian/changesets/issues/622
// @ts-expect-error
const assembleReleasePlan = (_assembleReleasePlan.default ||
  _assembleReleasePlan) as typeof _assembleReleasePlan
// @ts-expect-error
const parseChangeset = (_parseChangeset.default ||
  _parseChangeset) as typeof _parseChangeset

export const getChangedPackages = async ({
  changedFiles: changedFilesPromise,
}: {
  changedFiles: Promise<string[]> | string[]
  api: Gitlab
  // eslint-disable-next-line sonarjs/cognitive-complexity
}) => {
  let hasErrored = false
  function fetchFile(path: string) {
    return fs.readFile(path, 'utf8')
  }

  function fetchJsonFile<T = unknown>(path: string) {
    return fetchFile(path)
      .then(x => JSON.parse(x) as T)
      .catch((err: unknown) => {
        hasErrored = true
        console.error(err)
        return {} as unknown as T
      })
  }

  function fetchTextFile(path: string) {
    return fetchFile(path).catch((err: unknown) => {
      hasErrored = true
      console.error(err)
      return ''
    })
  }

  async function getPackage(pkgPath: string) {
    const jsonContent = await fetchJsonFile<PackageJSON>(
      pkgPath + '/package.json',
    )
    return {
      packageJson: jsonContent,
      dir: pkgPath,
    }
  }

  const rootPackageJsonContentsPromise = fetchJsonFile<
    PackageJSON & {
      workspaces?:
        | string[]
        | {
            packages: string[]
          }
      bolt?: {
        workspaces?: string[]
      }
    }
  >('package.json')
  const configPromise = fetchJsonFile<WrittenConfig>('.changeset/config.json')

  const tree = await getAllFiles(process.cwd())

  let preStatePromise: Promise<PreState> | undefined
  const changesetPromises: Array<Promise<NewChangeset>> = []
  const potentialWorkspaceDirectories: string[] = []
  let isPnpm = false
  const changedFiles = await changedFilesPromise

  for (const item of tree) {
    if (item.endsWith('/package.json')) {
      const dirPath = nodePath.dirname(item)
      potentialWorkspaceDirectories.push(dirPath)
    } else if (item === 'pnpm-workspace.yaml') {
      isPnpm = true
    } else if (item === '.changeset/pre.json') {
      preStatePromise = fetchJsonFile('.changeset/pre.json')
    } else if (
      item !== '.changeset/README.md' &&
      item.startsWith('.changeset') &&
      item.endsWith('.md') &&
      changedFiles.includes(item)
    ) {
      const res = /\.changeset\/([^.]+)\.md/.exec(item)
      if (!res) {
        throw new Error('could not get name from changeset filename')
      }
      const id = res[1]
      changesetPromises.push(
        fetchTextFile(item).then(text => ({
          ...parseChangeset(text),
          id,
        })),
      )
    }
  }
  let tool:
    | {
        tool: Tool
        globs: string[]
      }
    | undefined

  if (isPnpm) {
    tool = {
      tool: 'pnpm',
      globs: (
        load(await fetchTextFile('pnpm-workspace.yaml')) as {
          packages: string[]
        }
      ).packages,
    }
  } else {
    const rootPackageJsonContent = await rootPackageJsonContentsPromise

    if (rootPackageJsonContent.workspaces) {
      // eslint-disable-next-line no-negated-condition
      tool = !Array.isArray(rootPackageJsonContent.workspaces)
        ? {
            tool: 'yarn',
            globs: rootPackageJsonContent.workspaces.packages,
          }
        : {
            tool: 'yarn',
            globs: rootPackageJsonContent.workspaces,
          }
    } else if (rootPackageJsonContent.bolt?.workspaces) {
      tool = {
        tool: 'bolt',
        globs: rootPackageJsonContent.bolt.workspaces,
      }
    }
  }

  const rootPackageJsonContent = await rootPackageJsonContentsPromise

  const packages: Packages = {
    root: {
      dir: '/',
      packageJson: rootPackageJsonContent,
    },
    tool: tool ? tool.tool : 'root',
    packages: [],
  }

  if (tool) {
    if (
      !Array.isArray(tool.globs) ||
      !tool.globs.every(x => typeof x === 'string')
    ) {
      throw new Error('globs are not valid: ' + JSON.stringify(tool.globs))
    }
    const matches = micromatch(potentialWorkspaceDirectories, tool.globs)

    packages.packages = await Promise.all(matches.map(dir => getPackage(dir)))
  } else {
    packages.packages.push(packages.root)
  }
  // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition -- https://github.com/microsoft/TypeScript/issues/9998
  if (hasErrored) {
    throw new Error('an error occurred when fetching files')
  }

  const config = await configPromise.then(rawConfig =>
    parseConfig(rawConfig, packages),
  )

  const releasePlan = assembleReleasePlan(
    await Promise.all(changesetPromises),
    packages,
    config,
    await preStatePromise,
  )

  return {
    changedPackages: (packages.tool === 'root'
      ? packages.packages
      : packages.packages.filter(pkg =>
          changedFiles.some(changedFile => changedFile.includes(pkg.dir)),
        )
    )
      .filter(
        pkg =>
          pkg.packageJson.private !== true &&
          !config.ignore.includes(pkg.packageJson.name),
      )
      .map(x => x.packageJson.name),
    releasePlan,
  }
}
