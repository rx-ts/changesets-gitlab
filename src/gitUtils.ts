import { exec } from '@actions/exec'

import { execWithOutput, identify } from './utils.js'

export const setupUser = async () => {
  await exec('git', [
    'config',
    '--global',
    'user.name',
    process.env.GITLAB_USER_NAME!,
  ])
  await exec('git', [
    'config',
    '--global',
    'user.email',
    process.env.GITLAB_CI_USER_EMAIL ||
      '"gitlab[bot]@users.noreply.gitlab.com"',
  ])
}

export const pullBranch = async (branch: string) => {
  await exec('git', ['pull', 'origin', branch])
}

export const push = async (
  branch: string,
  { force }: { force?: boolean } = {},
) => {
  await exec(
    'git',
    ['push', 'origin', `HEAD:${branch}`, force && '--force'].filter(identify),
  )
}

export const pushTags = async () => {
  await exec('git', ['push', 'origin', '--tags'])
}

export const switchToMaybeExistingBranch = async (branch: string) => {
  const { stderr } = await execWithOutput('git', ['checkout', branch], {
    ignoreReturnCode: true,
  })
  const isCreatingBranch = !stderr
    .toString()
    .includes(`Switched to a new branch '${branch}'`)
  if (isCreatingBranch) {
    await exec('git', ['checkout', '-b', branch])
  }
}

export const reset = async (
  pathSpec: string,
  mode: 'hard' | 'mixed' | 'soft' = 'hard',
) => {
  await exec('git', ['reset', `--${mode}`, pathSpec])
}

export const commitAll = async (message: string) => {
  await exec('git', ['add', '.'])
  await exec('git', ['commit', '-m', message])
}

export const checkIfClean = async (): Promise<boolean> => {
  const { stdout } = await execWithOutput('git', ['status', '--porcelain'])
  return stdout.length === 0
}
