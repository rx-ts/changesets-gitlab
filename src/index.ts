import { Gitlab } from '@gitbeaker/rest'
import type { ProxyAgentConfigurationType } from 'global-agent'
import { bootstrap } from 'global-agent'

import { env } from './env.js'

const PROXY_PROPS = ['http_proxy', 'https_proxy', 'no_proxy'] as const

declare global {
  const GLOBAL_AGENT: {
    HTTP_PROXY: string | null
    HTTPS_PROXY: string | null
    NO_PROXY: string | null
  }
}

export const createApi = (gitlabToken?: string) => {
  bootstrap()

  for (const prop of PROXY_PROPS) {
    const uProp = prop.toUpperCase() as keyof ProxyAgentConfigurationType
    const value = process.env[uProp] || process.env[prop]
    if (value) {
      GLOBAL_AGENT[uProp] = value
    }
  }

  const token = gitlabToken || env.GITLAB_TOKEN
  const host = env.GITLAB_HOST

  // we cannot use { [tokenType]: token } now
  // because it will break the type of the Gitlab constructor
  switch (env.GITLAB_TOKEN_TYPE) {
    case 'job': {
      return new Gitlab({
        host,
        jobToken: token,
      })
    }
    case 'oauth': {
      return new Gitlab({
        host,
        oauthToken: token,
      })
    }
    default: {
      return new Gitlab({
        host,
        token,
      })
    }
  }
}
