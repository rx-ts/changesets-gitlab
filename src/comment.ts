import { ValidationError } from '@changesets/errors'
import type {
  ComprehensiveRelease,
  ReleasePlan,
  VersionType,
} from '@changesets/types'
import type { Gitlab } from '@gitbeaker/core'
import type { MergeRequestDiffSchema } from '@gitbeaker/rest'
import { captureException } from '@sentry/node'
import { humanId } from 'human-id'
import { markdownTable } from 'markdown-table'

import * as context from './context.js'
import { env } from './env.js'
import { getChangedPackages } from './get-changed-packages.js'
import { getUsername } from './utils.js'

import { createApi } from './index.js'

const generatedByBotNote = 'Generated By Changesets GitLab Bot'

const getReleasePlanMessage = (releasePlan: ReleasePlan | null) => {
  if (!releasePlan) return ''

  const publishableReleases = releasePlan.releases.filter(
    (x): x is ComprehensiveRelease & { type: Exclude<VersionType, 'none'> } =>
      x.type !== 'none',
  )

  const table = markdownTable([
    ['Name', 'Type'],
    ...publishableReleases.map(x => [
      x.name,
      {
        major: 'Major',
        minor: 'Minor',
        patch: 'Patch',
      }[x.type],
    ]),
  ])

  return `<details><summary>This MR includes ${
    releasePlan.changesets.length > 0
      ? `changesets to release ${
          publishableReleases.length === 1
            ? '1 package'
            : `${publishableReleases.length} packages`
        }`
      : 'no changesets'
  }</summary>

  ${
    publishableReleases.length > 0
      ? table
      : "When changesets are added to this MR, you'll see the packages that this MR includes changesets for and the associated semver types"
  }

</details>`
}

const getAbsentMessage = (
  commitSha: string,
  addChangesetUrl: string,
  releasePlan: ReleasePlan | null,
) => `###  ⚠️  No Changeset found

Latest commit: ${commitSha}

Merging this MR will not cause a version bump for any packages. If these changes should not result in a new version, you're good to go. **If these changes should result in a version bump, you need to add a changeset.**

${getReleasePlanMessage(releasePlan)}

[Click here to learn what changesets are, and how to add one](https://github.com/changesets/changesets/blob/master/docs/adding-a-changeset.md).

[Click here if you're a maintainer who wants to add a changeset to this MR](${addChangesetUrl})

__${generatedByBotNote}__
`

const getApproveMessage = (
  commitSha: string,
  addChangesetUrl: string,
  releasePlan: ReleasePlan | null,
) => `###  🦋  Changeset detected

Latest commit: ${commitSha}

**The changes in this MR will be included in the next version bump.**

${getReleasePlanMessage(releasePlan)}

Not sure what this means? [Click here  to learn what changesets are](https://github.com/changesets/changesets/blob/master/docs/adding-a-changeset.md).

[Click here if you're a maintainer who wants to add another changeset to this MR](${addChangesetUrl})

__${generatedByBotNote}__
`

const getNewChangesetTemplate = (changedPackages: string[], title: string) =>
  encodeURIComponent(`---
${changedPackages.map(x => `"${x}": patch`).join('\n')}
---

${title}
`)

const getNoteInfo = (api: Gitlab, mrIid: number | string) =>
  api.MergeRequestDiscussions.all(context.projectId, mrIid).then(
    async discussions => {
      for (const discussion of discussions) {
        if (!discussion.notes) {
          continue
        }

        const username = await getUsername(api)
        const changesetBotNote = discussion.notes.find(
          note =>
            note.author.username === username &&
            // We need to ensure the note is generated by us, but we don't have an app bot like GitHub
            // @see https://github.com/apps/changeset-bot
            note.body.includes(generatedByBotNote),
        )

        if (changesetBotNote) {
          return {
            discussionId: discussion.id,
            noteId: changesetBotNote.id,
          }
        }
      }
    },
  )

const hasChangesetBeenAdded = async (
  allDiffsPromise: Promise<MergeRequestDiffSchema[]>,
) => {
  const allDiffs = await allDiffsPromise
  return allDiffs.some(file => {
    return (
      file.new_file &&
      /^\.changeset\/.+\.md$/.test(file.new_path) &&
      file.new_path !== '.changeset/README.md'
    )
  })
}

export const comment = async () => {
  const mrBranch = env.CI_MERGE_REQUEST_SOURCE_BRANCH_NAME
  if (!mrBranch) {
    console.warn('[changesets-gitlab:comment] It should only be used on MR')
    return
  }

  const {
    CI_MERGE_REQUEST_IID: mrIid,
    CI_MERGE_REQUEST_PROJECT_URL,
    CI_MERGE_REQUEST_SOURCE_BRANCH_SHA,
    CI_MERGE_REQUEST_TITLE,
    GITLAB_COMMENT_TYPE,
    GITLAB_ADD_CHANGESET_MESSAGE,
  } = env

  if (mrBranch.startsWith('changeset-release')) {
    return
  }

  const api = createApi()

  let errFromFetchingChangedFiles = ''
  try {
    const latestCommitSha = CI_MERGE_REQUEST_SOURCE_BRANCH_SHA
    const allDiffsPromise = api.MergeRequests.allDiffs(context.projectId, mrIid)

    const [noteInfo, hasChangeset, { changedPackages, releasePlan }] =
      await Promise.all([
        getNoteInfo(api, mrIid),
        hasChangesetBeenAdded(allDiffsPromise),
        getChangedPackages({
          changedFiles: allDiffsPromise.then(x => x.map(x => x.new_path)),
          api,
        }).catch((err: unknown) => {
          if (err instanceof ValidationError) {
            errFromFetchingChangedFiles = `<details><summary>💥 An error occurred when fetching the changed packages and changesets in this MR</summary>\n\n\`\`\`\n${err.message}\n\`\`\`\n\n</details>\n`
          } else {
            console.error(err)
            captureException(err)
          }
          return {
            changedPackages: ['@fake-scope/fake-pkg'],
            releasePlan: null,
          }
        }),
      ] as const)

    const addChangesetUrl = `${CI_MERGE_REQUEST_PROJECT_URL}/-/new/${mrBranch}?file_name=.changeset/${humanId(
      {
        separator: '-',
        capitalize: false,
      },
    )}.md&file=${getNewChangesetTemplate(
      changedPackages,
      CI_MERGE_REQUEST_TITLE,
    )}${
      GITLAB_ADD_CHANGESET_MESSAGE
        ? '&commit_message=' + encodeURIComponent(GITLAB_ADD_CHANGESET_MESSAGE)
        : ''
    }`

    const prComment =
      (hasChangeset
        ? getApproveMessage(latestCommitSha, addChangesetUrl, releasePlan)
        : getAbsentMessage(latestCommitSha, addChangesetUrl, releasePlan)) +
      errFromFetchingChangedFiles

    if (GITLAB_COMMENT_TYPE === 'discussion') {
      if (noteInfo) {
        return api.MergeRequestDiscussions.editNote(
          context.projectId,
          mrIid,
          noteInfo.discussionId,
          noteInfo.noteId,
          {
            body: prComment,
          },
        )
      }

      return api.MergeRequestDiscussions.create(
        context.projectId,
        mrIid,
        prComment,
      )
    }

    if (GITLAB_COMMENT_TYPE === 'note') {
      if (noteInfo) {
        return api.MergeRequestNotes.edit(
          context.projectId,
          mrIid,
          noteInfo.noteId,
          { body: prComment },
        )
      }

      return api.MergeRequestNotes.create(context.projectId, mrIid, prComment)
    }

    throw new Error(
      `Invalid comment type "${GITLAB_COMMENT_TYPE}", should be "discussion" or "note"`,
    )
  } catch (err: unknown) {
    console.error(err)
    throw err
  }
}
