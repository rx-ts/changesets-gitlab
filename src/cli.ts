#!/usr/bin/env node

import './env.js'

import _ from 'node:module'

import { program } from 'commander'

import { comment } from './comment.js'
import { main } from './main.js'
import { getOptionalInput } from './utils.js'

const cjsRequire =
  typeof require === 'undefined' ? _.createRequire(import.meta.url) : require

const run = async () => {
  program.version(
    (cjsRequire('../package.json') as { version: string }).version,
  )

  program.command('comment').action(async () => {
    await comment()
  })

  program
    .command('main', {
      isDefault: true,
    })
    .action(() =>
      main({
        published: getOptionalInput('published'),
        onlyChangesets: getOptionalInput('only_changesets'),
      }),
    )

  return program.showHelpAfterError().parseAsync()
}

run().catch((err: Error) => {
  console.error(err)
  process.exitCode = 1
})
