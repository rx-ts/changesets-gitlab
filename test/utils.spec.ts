import { getAllFiles } from '../src/utils'

test('utils', async () => {
  expect(await getAllFiles('test/fixtures')).toMatchSnapshot()
})
