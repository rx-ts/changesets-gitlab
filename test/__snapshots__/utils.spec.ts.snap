// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`utils getAllFiles 1`] = `
[
  ".gitkeep",
  "dir/.gitkeep",
  "dir/subdir/.gitkeep",
]
`;
